### ISSUE TITLE
Import and render subfolders

### ISSUE THREAD

**Issue by juliangruber** [hypergraph-xyz/desktop#110]
> This would work and I think is the best approach for now if adding folders adds significant dev time, but I don't see why we shouldn't allow adding of entire folders at once. This at least offers some easy way of keeping a folder structure intact and adding many files at once, without complex logic regarding what goes where. My suggestion would be this:
> 
> * allow files and folders to be added
> * every selected file or folder will be put into the archive's root
> * the only subfolders that are created are created when the user selects an entire folder to add
> * if there are conflicts between individually added files or folders, change the names to remove conflicts, eg
>   
>   * `index.html`
>   * `index (2).html`
> 
> In short: treat a selected entire folder in the same way that you would treat one individual file.

> I have removed adding folders again because our files design doesn't allow for deleting folders, only individual files, so mistakes would be hard to recover from.

**Comment by juliangruber** [comment]
@jameslibscie @chartgerink I'm thinking about how to implement this, and might have a solution.

The reason this isn't trivial is that say you select five files:

- `~/project-a/index.html`
- `~/project-a/index.js`
- `~/project-b/index.html`
- `~/project-b/index.js`
- `~/some-video.mp4`

How could the app determine where to place which file? Looking at this example, finding a folder that contains all of the files (=lowest common denominator) would work. Then the files would be placed at

- `ROOT/project-a/index.html`
- `ROOT/project-a/index.js`
- `ROOT/project-b/index.html`
- `ROOT/project-b/index.js`
- `ROOT/some-video.mp4`

While this might not be exactly what the user wants, there are no conflicts and the original structure is kept.

However, look at a different scenario, where the user wants to add these two files:

- `~/dev/juliangruber/project/findings.csv`
- `~/Downloads/report.pdf`

Placing them at..

- `ROOT/dev/juliangruber/project/findings.csv`
- `ROOT/Downlodas/report.pdf`

..is most likely not what they want. Instead, they probably would have wanted

- `ROOT/findings.csv`
- `ROOT/report.pdf`

Now, saying only to use the lowest common denominator strategy when there is conflicts in file names is confusing and unpredictable.

I've therefore looked at how we implemented this in Dat Desktop, and discovered:

https://github.com/datproject/dat-desktop/blob/474ee9dcc5bae8c915ecb2c7e9c7aaf9f0783820/app/actions/index.js#L31-L34

Dat Desktop only allows the user to select one folder. Not multiple folders, not multiple files, not a single file. Only one folder, which will then become the module folder.

This has none of the problems mentioned above, however it is less convenient: If the user doesn't already have a folder prepared, they need to create a fake one for the module, only for the files to be copied over to the p2pcommons folder anyway.

I'm curious what you two think about this. Should we wait with this until we have richer file management, with folders, drag & drop, etc? Should we go any of the ways mentioned above?

**Comment by jameslibscie** [comment]
I agree with your analysis. I don't think selecting a single folder is acceptable, as this will likely lead to modules containing basically entire papers / projects. Perhaps adding folders is something that can wait until drag 'n' drop, but I have two other ideas.

1. Is there not a way for the OS "select file" dialog to accept folders as input, rather than just files? I'm pretty sure I've seen this before.

2. What if we conditionally deduce the folder structure depending on the relationship between files? In your first example, there is a file located in the root folder which has also been selected. This indicates that the folder structure is relevant. In your second example, there isn't, indicating that the folder structure isn't relevant. I guess the logic would be finding a lowest common denominator in the paths and then checking whether there is a file located there.

An example where this would go wrong, but is unlikely to happen I think:
- `C:\somedoc.html`
- `C:\Users\James\Documents\Research\project-a\data.csv`

This might lead to people not understanding the logic and therefore ending up with other results than expected. The ideal thing would be to suggest a folder structure that people can easily edit, but this doesn't seem easy and also feels like over-optimisation right now.

**Comment by juliangruber** [comment]
1. Absolutely, this is possible. I don’t see how this solves any cases above however, since it doesn’t make a difference whether you select all files and folders at once or one after the other.

**Comment by juliangruber** [comment]
2. Lowest common denominator + Checking whether there is a file there is also too complicated in my understanding. It will behave unexpectedly and users will need explanation.

**Comment by juliangruber** [comment]
If we only let people select a single folder none of these edge cases exist, however it will require them to prepare a folder before adding it. A bit more work but it is reliable and understandable.

**Comment by chartgerink** [comment]
How about we leave the importing of folders for now, as we do ask people to import files still, and only adjust the rendering of the path? 

That way if somebody adds a subfolder locally, it will display, but we don't necessarily create any additional overhead in prepping the import. We can spend the time instead on integrating drag-and-drop of files.

**Comment by jameslibscie** [comment]
> 1. Absolutely, this is possible. I don’t see how this solves any cases above however, since it doesn’t make a difference whether you select all files and folders at once or one after the other.

If you select files `main.html`, `extra.html` and folder `folder1`, you would get this:
- `ROOT/main.html`
- `ROOT/extra.html`
- `ROOT/folder1/...files`

If you select files `main.html`, `extra.html` and add those, then go into `folder1` and add the files there, you would get this:
- `ROOT/main.html`
- `ROOT/extra.html`
- `ROOT/...files`

So whatever you select goes directly into `ROOT`. The ability to select a folder in addition to files seems like a big improvement to me as it allows you to add a whole folder at once AND keep folder structure.


> 2. Lowest common denominator + Checking whether there is a file there is also too complicated in my understanding. It will behave unexpectedly and users will need explanation.

Cool, I can definitely imagine this being too complicated


> How about we leave the importing of folders for now, as we do ask people to import files still, and only adjust the rendering of the path?
> 
> That way if somebody adds a subfolder locally, it will display, but we don't necessarily create any additional overhead in prepping the import. We can spend the time instead on integrating drag-and-drop of files.

👍

**Comment by juliangruber** [comment]
> How about we leave the importing of folders for now, as we do ask people to import files still, and only adjust the rendering of the path?

The tough part is adjusting how to display the path, not how to import folders. All the edge cases above still remain.

**Comment by juliangruber** [comment]
> If you select files `main.html`, `extra.html` and add those, then go into `folder1` and add the files there, you would get this:
> 
> * `ROOT/main.html`
> * `ROOT/extra.html`
> * `ROOT/...files`
> 
> So whatever you select goes directly into `ROOT`.

This is already the current behavior and doesn't solve cases like importing `folder1`/index.html` and `folder2/index.html`.

**Comment by juliangruber** [comment]
So for except for only letting the user add one folder per module there's no suggestion in this thread (as far as I understand) that can work correctly.

**Comment by juliangruber** [comment]
Outcome of @chartgerink and me talking about this:

- as is currently, only allow files to be added
- as is currently, allow multiple files to be added at once, but not folders
- every file will be put into the archive's root, there are no subfolders
- if there are conflicts between the file names, change the names to remove conflicts, eg
  - `index.html`
  - `index (2).html`

This is straight forward to implement and should work for now.

@jameslibscie what do you think?

**Comment by jameslibscie** [comment]
This would work and I think is the best approach for now if adding folders adds significant dev time, but I don't see why we shouldn't allow adding of entire folders at once. This at least offers some easy way of keeping a folder structure intact and adding many files at once, without complex logic regarding what goes where. My suggestion would be this:

- allow files and folders to be added
- every selected file or folder will be put into the archive's root
- the only subfolders that are created are created when the user selects an entire folder to add
- if there are conflicts between individually added files or folders, change the names to remove conflicts, eg
  - `index.html`
  - `index 2.html`

In short: treat a selected entire folder in the same way that you would treat one individual file.

**Comment by juliangruber** [comment]
Cool, @jameslibscie that sounds good to me! No need to keep that one constraint.

**Comment by juliangruber** [comment]
I have removed adding folders again because our files design doesn't allow for deleting folders, only individual files, so mistakes would be hard to recover from.

**Comment by jameslibscie** [comment]
> I have removed adding folders again because our files design doesn't allow for deleting folders, only individual files, so mistakes would be hard to recover from.

If I manually add a folder by pasting it into the module folder, it shows in Hypergraph as if it were a single file. Could we not handle the adding of folders through Hypergraph in the same way? So we can add whole folders, open whole folders and remove whole folders (when the edit content issue is done), but not do anything within those folders within Hypergraph?

Just a suggestion, if this is significant work now, then let's leave it as-is. Let me know and I'll approve.

**Comment by juliangruber** [comment]
> If I manually add a folder by pasting it into the module folder, it shows in Hypergraph as if it were a single file.

I would consider this a bug, why would we list files of the root folder but not of a sub folder? We can implement a more consistent behavior now when not allowing folders.
